File: apps/frontend/src/types/automations.ts
export type WorkflowStatus = "DRAFT" | "PUBLISHED" | "DISABLED"
export type ExecutionStatus = "RUNNING" | "SUCCESS" | "FAILED" | "CANCELED"
export type TriggerType = "manual" | "webhook" | "system"
export type NodeCategory = "trigger" | "action" | "logic" | "utils" | "transform"

export type NodeType =
    | "ManualTrigger" | "WebhookTrigger"
    | "HttpRequest" | "SendEmail" | "CreateCRMTask"
    | "IfCondition" | "Delay"
    | "SetVariable" | "Transform"
    | "CodeScript" | "FilterItems" | "SortItems" | "RemoveDuplicates"
    | "SplitBatches" | "MergeData" | "DateTimeFormat" | "RenameKeys"

export interface NodePosition { x: number; y: number }

export interface WorkflowNode {
    id: string
    type: NodeType
    category: NodeCategory
    position: NodePosition
    data: { label: string; config: Record<string, unknown> }
}

export interface WorkflowEdge {
    id: string
    source: string
    target: string
    sourceHandle?: string
    targetHandle?: string
    label?: string
}

export interface WorkflowViewport { x: number; y: number; zoom: number }

export interface WorkflowDefinition {
    schemaVersion: "1.0"
    meta: { name: string; description?: string; tags?: string[]; createdAt?: string; updatedAt?: string }
    viewport: WorkflowViewport
    nodes: WorkflowNode[]
    edges: WorkflowEdge[]
}

export interface WorkflowResponse {
    id: string
    organization_id: string
    name: string
    description: string | null
    status: WorkflowStatus
    current_version_id: string | null
    created_by: string | null
    updated_by: string | null
    created_at: string
    updated_at: string
}

export interface WorkflowVersionResponse {
    id: string
    workflow_id: string
    version_number: number
    definition_json: WorkflowDefinition
    published_at: string | null
    created_at: string
}

export interface ExecutionResponse {
    id: string
    workflow_id: string
    version_id: string | null
    status: ExecutionStatus
    trigger_type: TriggerType
    started_at: string
    finished_at: string | null
    duration_ms: number | null
    initiated_by: string | null
    error_summary: string | null
}

export interface ExecutionStepResponse {
    id: string
    execution_id: string
    node_id: string
    node_type: string
    status: string
    started_at: string
    finished_at: string | null
    duration_ms: number | null
    input_json: Record<string, unknown> | null
    output_json: Record<string, unknown> | null
    error_message: string | null
}

export interface WorkflowCreateInput { name: string; description?: string; definition?: WorkflowDefinition }
export interface WorkflowUpdateInput { name?: string; description?: string; definition?: WorkflowDefinition }
export interface PublishInput { definition: WorkflowDefinition }
export interface ManualExecuteInput { payload?: Record<string, unknown> }
export interface WorkflowListParams { status?: WorkflowStatus; limit?: number; offset?: number }
export interface ExecutionListParams { workflow_id?: string; status?: ExecutionStatus; limit?: number; offset?: number }

File: apps/backend/src/schemas/automations.py
"""Schemas Pydantic v2 para o módulo Automations."""
from datetime import datetime
from typing import Any, Optional
from pydantic import BaseModel, ConfigDict, Field


class WorkflowCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    definition: Optional[dict[str, Any]] = None


class WorkflowUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    definition: Optional[dict[str, Any]] = None


class WorkflowResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    organization_id: str
    name: str
    description: Optional[str]
    status: str
    current_version_id: Optional[str]
    created_by: Optional[str]
    updated_by: Optional[str]
    created_at: datetime
    updated_at: datetime


class WorkflowVersionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    workflow_id: str
    version_number: int
    definition_json: dict[str, Any]
    published_at: Optional[datetime]
    created_at: datetime


class PublishBody(BaseModel):
    definition: dict[str, Any]


class ExecutionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    workflow_id: str
    version_id: Optional[str]
    status: str
    trigger_type: str
    started_at: datetime
    finished_at: Optional[datetime]
    duration_ms: Optional[int]
    initiated_by: Optional[str]
    error_summary: Optional[str]


class ExecutionStepResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    execution_id: str
    node_id: str
    node_type: str
    status: str
    started_at: datetime
    finished_at: Optional[datetime]
    duration_ms: Optional[int]
    input_json: Optional[dict[str, Any]]
    output_json: Optional[dict[str, Any]]
    error_message: Optional[str]


class WebhookTriggerBody(BaseModel):
    model_config = ConfigDict(extra="allow")


class ManualExecuteBody(BaseModel):
    payload: Optional[dict[str, Any]] = Field(default_factory=dict)
